File: 2-Sorting/6-heap-sort.cpp
#include <bits/stdc++.h>
using namespace std;

int hepify(vector<int>& arr, int n, int node)  { 
    int left = 2 * node + 1;
    int right = 2 * node + 2;

    int root = node;

    if (left < n && arr[left] > arr[root]) {
        root = left;
    }

    if (right < n && arr[right] > arr[root]) {
        root = right;
    }

    if (root != node) {
        swap(arr[node], arr[root]);
        hepify(arr, n, root);
    }
} 

void heap_sort(vector<int>& arr) {
    for (int i = arr.size() / 2 - 1; i >= 0; i--) {
        hepify(arr, arr.size(), i);
    }

    for (int i = arr.size() - 1; i >= 0; i--) {
        swap(arr[i], arr[0]);
        hepify(arr, i, 0);
    }
}

void test() {
    int n;
    cin >> n;

    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    heap_sort(arr);

    for (int i : arr) {
        cerr << i << " ";
    }
    cerr << "\n";
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    srand(time(nullptr));

    test();

    return 0;
}

File: 2-Sorting/rust/6-heap-sort.rs
// Heap sort

mod sort {
    /// Sort the given array via heap sort
    /// Complexity: O(array.len()*log2(array.len()))
    pub fn heap_sort(array: &mut Vec<i32>) {
        for i in (0..array.len() / 2).rev() {
            hepify(array, array.len(), i);
        }
        
        for i in (0..array.len()).rev() {
            array.swap(i, 0);
            hepify(array, i, 0);
            
        }
    }
    
    fn hepify(array: &mut Vec<i32>, n: usize, node: usize) {
        let left = 2 * node + 1;
        let right = 2 * node + 2;
    
        let mut new_node = node;
    
        if left < n && array[left] > array[new_node] {
            new_node = left;
        }
    
        if right < n && array[right] > array[new_node] {
            new_node = right;
        }
    
        if new_node != node {
            array.swap(node, new_node);
            hepify(array, n, new_node);
        }
    }
}

#[cfg(test)]
mod tests {

    #[test]
    fn test_heap_sort() {
        let mut test_array: Vec<i32> = vec![5, 1, 4, 7, 13, 2];
        let expected_result: Vec<i32> = vec![1, 2, 4, 5, 7, 13];
    
        crate::sort::heap_sort(&mut test_array);
    
        assert_eq!(expected_result, test_array);
    }
}
